### ISSUE TITLE
RFC Add API to do post file upload processing

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/silverstripe-assets#399]
We have an ongoing problems where when you create new file outside of asset-admin, you have to manually create thumbnails for them. So we keep copying the same bit of code everywhere so we can generate asset-admin thumbnails and we're still finding spots where this is missing. e.g. https://github.com/silverstripe/silverstripe-userforms/issues/968

Another example could be where you need to index the files after they've been uploaded, or add automatic tags to them.

This is kind of ridiculous. When creating files, you shouldn't have to manually generate thumbnails or worry about other modules that might need to do some post-processing. A better way would be to have one place where you can register some task that need to be run after each file upload.

Than asset-admin could register a task to run after each file upload to generate its thumbnails. Even better, you could configure this task to work with queued job so if you expect the it to take some time, you can defer it until later.

**Comment by dhensby** [comment]
Yes, sounds like a good idea. Some kind of file upload handler where you can register middlewares or file upload event?

The lack of an event system in SS makes the second approach difficult, but it's been something I think has been missing for quite some time so could be added at both times. [PSR-14 is relevant here](https://www.php-fig.org/psr/psr-14/)

**Comment by emteknetnz** [comment]
If we were to limit this strictly to file uploads, then [FileField::saveInto()](https://github.com/silverstripe/silverstripe-framework/blob/4/src/Forms/FileField.php#L129) may play a part since that's what's used in both the front and back end to upload files

Otherwise if it were literally whenever a File is created then `File::create()` ?

> register middlewares or file upload event?

Without getting fancy, given that it's a core module, good old extension hooks would work OK here?

**Comment by maxime-rainville** [comment]
My instinct would be to do it when a new File is created or when its content is updated rather than do it on upload.

Otherwise if you programmatically create a file, that logic won't fire.

**Comment by emteknetnz** [comment]
That makes sense, if we were to limit this to just thumbnail creation then `Image::create()` and extension-hooks seems pretty good

Would increasing the scope to having some sort of fancy abstract way to handle "stuff happening when you create a file" be worth doing? Or is this just overthinking things?

**Comment by polycode-nz** [comment]
I'm having this issue when doing manual uploads from a front end form.

When looping through $_FILES array I'm using `AssetAdmin::singleton()->generateThumbnails($file);` which is showing a blue square in the admin assets area.

Truncated example inside $_FILES loop:

```
$file = File::create();
$upload = Upload::create();
$upload->loadIntoFile($attachment, $file, '/Submission attachments');

$file->write();

$test = Attachment::create();
$test->Title = $attachment['name'];
$test->FileID = $upload->getFile()->ID;

$submission->Attachment()->add($test);

AssetAdmin::singleton()->generateThumbnails($file);
```

Is this expected to generate a thumbnail or am I doing something wrong?

PDF files do get the correct thumbnail.

**Comment by maxime-rainville** [comment]
@polycode-nz This RFC is more about how we should approach this issue at an abstract high level, while it sounds like you are looking for a fix to a immediate problem. Would you mind starting a [forum thread](https://forum.silverstripe.org/) and explain the problem you are facing there? I would rather keep the discussion here focus on the RFC.

Just flag me into your forum post and I'll try my best to help you.
